### ISSUE TITLE
What's slow about this algorithm in Sidef? 

### ISSUE THREAD

**Issue by catb0t** [trizen/sidef#37]
The game is to write a toy digit-by-digit common logarithm (log10) function; I'm using Sidef's arbitrary precision to test the algorithm on non-hardware numbers.

The algorithm is from http://brics.dk/RS/04/17/BRICS-RS-04-17.pdf, but the algorithm isn't hugely important here.

The function in Sidef is 

```perl
func log10_ (x, g) {
  const times = g*x.to_s.len
  var k = x
  gather {
    for (^ times) {
      const s = k.to_s
      const d = (s.chars.slice(1, s.index(".")).len - 1)
      take d
      k = ((k / (10**d)) ** 10)
      say k
    }
  }
}

say log10(1234.56)
say "---"
say log10_(1234.56, 2) # getting 2*6 digits of the logarithm
```

The equivalent code in Python is 

```python
def log10_1p(x, g):
    logdigits = []
    s = str(x)
    x = float(x)
    for _ in range(g * len(s)):
        s = str(x)
        logdigits.append(len(s[1 : s.index(".") or len(s) - 1]))
        x = (x / (10 ** logdigits[-1])) ** 10
        print(x)
    return logdigits

print(log10_1p(1234.56, 3)) # getting 3*6 digits of the logarithm
```

The output from Python is 
```
8.22473693827674
1416511689.4062471
32.52382591129096
132439.11735410485
16.602090509224485
159.08446495477853
103.8192091772769
[...]
[3, 0, 9, 1, 5, 1, 2, 2, 0, 1, 6, 2, 7, 7]

real	0m0.062s
```

When running the Sidef code the output is 
```
3.0915122016278
---
8.22473693827675
1416511689.4062604015717
32.523825911294
132439.117354229694
16.602090509381
159.0844649697746
[...]
```
The `k` and `x` intermediate values are the same from Python and Sidef so the Sidef code must be a correct version of the algorithm.

After `159`, the Sidef code slows down, allocates almost 6GB of memory and Perl crashes -- the program never completes. 

There are two main implementation differences between CPython and Sidef. 

1. Python strings are immutable, Sidef Strings are (I think) mutable. This might mean Sidef is less efficient.

2. Python (usually) uses hardware floats or a very similar implementation, not GMP-style bignums. This is probably where most of the slowdown comes from.

Basically I just want to know, is it a bug that what seems to be very simple code is very slow, or am I mis-using Sidef?

I'm using `Sidef-3.17/bin/sidef -O2 -c -o logs.pl logs.sidef && perl logs.pl` to compile (`-P` option has no effect on speed) and `perl 5, version 26, subversion 1 (v5.26.1) built for x86_64-linux-gnu-thread-multi`.

**Comment by trizen** [comment]
Hi! Thanks for using Sidef. :)

By default, Sidef uses rational numbers (unlike python), therefore the value of `k` becomes extremely large.

The trick to get around this, is to convert `k` to floating-point number:

```ruby
k = float((k / (10**d)) ** 10)
```

Or, alternatively, create `k` to be a floating-point number in the first place:

```ruby
var k = x.float
```

**Comment by catb0t** [comment]
Excellent, thanks! By the way, is there any other way than reading the source code to find the list of methods defined on an object? 

In Python I'd use `dir(x)` and it says `['__abs__', '__add__', '__bool__', '__class__', ...]`, but I haven't found anything similar in Sidef?

**Comment by trizen** [comment]
You're welcome!

To get the list of available methods on an object, one can call the `.methods` method on an object:

```ruby
say 42.methods.keys.sort.join("\n")
```

Also, please see: https://github.com/trizen/sidef/wiki#built-in-classes

By clicking on a class, it will display the methods defined inside it (without documentation, for now).

For example: https://github.com/trizen/sidef/blob/master/lib/Sidef/Types/Number/Number.pod
